_posts/2021-02-11-tabnet/tabnet.Rmd

```diff
@@ -0,0 +1,411 @@
+
+```{r}
+grid <-
+  wf %>%
+  parameters() %>%
+  update(
+    decision_width = decision_width(range = c(20, 40)),
+    attention_width = attention_width(range = c(20, 40)),
+    num_steps = num_steps(range = c(4, 6)),
+    learn_rate = learn_rate(range = c(-2.5, -1))
+  ) %>%
+  grid_max_entropy(size = 8)
+
+grid
+```
+
+    # A tibble: 8 x 4
+      learn_rate decision_width attention_width num_steps
+           <dbl>          <int>           <int>     <int>
+    1    0.00529             28              25         5
+    2    0.0858              24              34         5
+    3    0.0230              38              36         4
+    4    0.0968              27              23         6
+    5    0.0825              26              30         4
+    6    0.0286              36              25         5
+    7    0.0230              31              37         5
+    8    0.00341             39              23         5
+
+To search the space, we use `tune_race_anova()` from the new [finetune](https://www.tidyverse.org/blog/2020/12/finetune-0-0-1/) package, making use of five-fold cross-validation:
+
+```{r}
+ctrl <- control_race(verbose_elim = TRUE)
+folds <- vfold_cv(train, v = 5)
+set.seed(777)
+
+res <- wf %>% 
+    tune_race_anova(
+    resamples = folds, 
+    grid = grid,
+    control = ctrl
+  )
+
+```
+
+We can now extract the best hyperparameter combinations:
+
+```{r}
+res %>% show_best("accuracy") %>% select(- c(.estimator, .config))
+```
+
+    # A tibble: 5 x 8
+      learn_rate decision_width attention_width num_steps .metric   mean     n std_err
+           <dbl>          <int>           <int>     <int> <chr>    <dbl> <int>   <dbl>
+    1     0.0858             24              34         5 accuracy 0.516     5 0.00370
+    2     0.0230             38              36         4 accuracy 0.510     5 0.00786
+    3     0.0230             31              37         5 accuracy 0.510     5 0.00601
+    4     0.0286             36              25         5 accuracy 0.510     5 0.0136 
+    5     0.0968             27              23         6 accuracy 0.498     5 0.00835
+
+It's hard to imagine how tuning could be more convenient!
+
+Now, we circle back to the original training workflow, and inspect TabNet's interpretability features.
+
+# TabNet interpretability features
+
+TabNet's most prominent characteristic is the way -- inspired by decision trees -- it executes in distinct steps. At each step, it again looks at the original input features, and decides which of those to consider based on lessons learned in prior steps. Concretely, it uses an attention mechanism to learn sparse *masks* which are then applied to the features.
+
+Now, these masks being "just" model weights means we can extract them and draw conclusions about feature importance. Depending on how we proceed, we can either
+
+-   aggregate mask weights over steps, resulting in global per-feature importances
```

Comment:
these bullets should be semi-colon delimited, and have an "or" before the last one (since they're not capitalized).

Reply:
OK!